src/darwin/Framework/CHIPTests/MTROTAProviderTests.m

```diff
@@ -1579,6 +1581,86 @@ - (void)test008_TestWriteDefaultOTAProviders
     }
 }
 
+#ifdef ENABLE_REAL_OTA_UPDATE_TESTS
+- (void)test009_TestOTAMetrics
+{
+    NSString * otaRawImagePath = [self createRawImageWithVersion:kUpdatedSoftwareVersion_5];
+    NSString * otaImagePath = [otaRawImagePath stringByReplacingOccurrencesOfString:@"raw-image" withString:@"image"];
+
+    // Check whether the ota raw image exists at otaRawImagePath
+    XCTAssertTrue([[NSFileManager defaultManager] fileExistsAtPath:otaRawImagePath]);
+
+    __auto_type * runner = [[MTROTARequestorAppRunner alloc] initWithPayload:kOnboardingPayload1 testcase:self];
+    __auto_type * device = [runner commissionWithNodeID:@(kDeviceId1)];
+
+    dispatch_queue_t queue = dispatch_get_main_queue();
+
+    __auto_type * delegate = [[MTRDeviceTestDelegate alloc] init];
+
+    __auto_type * checker =
+        [[MTROTAProviderTransferChecker alloc] initWithRawImagePath:otaRawImagePath
+                                           otaImageDownloadFilePath:runner.downloadFilePath
+                                                             nodeID:@(kDeviceId1)
+                                                    softwareVersion:kUpdatedSoftwareVersion_5
+                                              softwareVersionString:kUpdatedSoftwareVersionString_5
+                                                  applyUpdateAction:MTROTASoftwareUpdateProviderApplyUpdateActionProceed
+                                                           testcase:self];
+
+    // For some reason we need this to make sure the device in MTROTAImageTransferHandler::InvokeTransferSessionEndCallback
+    // has valid vendor id, product id, etc.
+    [device setDelegate:delegate queue:queue];
+
+    BDXTransferEndHandler existingTransferEndHandler = sOTAProviderDelegate.transferEndHandler;
+    sOTAProviderDelegate.transferEndHandler = ^(NSNumber * nodeID, MTRDeviceController * controller, MTRMetrics * metrics, NSError * _Nullable error) {
+        XCTAssertNil(error);
+
+        NSError * fileError = nil;
+        NSDictionary * fileAttributes = [[NSFileManager defaultManager] attributesOfItemAtPath:otaImagePath error:&fileError];
+        XCTAssertNil(fileError);
+
+        NSLog(@"%@", metrics);
+        uint32_t (^getValue)(NSString *, NSString *) = ^uint32_t(NSString * dataKey, NSString * valueKey) {
+            MTRMetricData * data = [metrics metricDataForKey:dataKey];
+            id value = [data valueForKey:valueKey];
+            XCTAssertNotNil(value);
+            NSLog(@"%@", value);
+            return [value unsignedIntValue];
+        };
+
+        XCTAssertEqual(getValue(@"dwnfw__ota__device_vendor_id_event", @"value"), kTestVendorId);
+        XCTAssertEqual(getValue(@"dwnfw__ota__device_product_id_event", @"value"), 32769);
```

Comment:
We should just add a `kTestProductId` which is `0x8001`, to match how we create our OTA images above.